pkg/compactor/blocks_cleaner.go

```diff
@@ -246,7 +270,15 @@ func (c *BlocksCleaner) cleanUpActiveUsers(ctx context.Context, users []string,
 	return concurrency.ForEachUser(ctx, users, c.cfg.CleanupConcurrency, func(ctx context.Context, userID string) error {
 		userLogger := util_log.WithUserID(userID, c.logger)
 		userBucket := bucket.NewUserBucketClient(userID, c.bucketClient, c.cfgProvider)
+		visitMarkerManager, err := c.obtainVisitMarkerManager(ctx, userLogger, userBucket)
+		if err != nil {
+			return err
+		}
+		if visitMarkerManager == nil {
+			return nil
+		}
 		errChan := make(chan error, 1)
+		go visitMarkerManager.HeartBeat(ctx, errChan, c.cleanerVisitMarkerFileUpdateInterval, true)
```

Comment:
Compaction logic would use visit marker later on. Thanos compact would take in errChan and push error if compaction failed. This would be only way to terminate heart beat in this case. For cleaner, it can do what you suggested but it cannot be reused for compaction logic.